public/app/features/manage-dashboards/state/actions.ts

```diff
@@ -45,17 +46,83 @@ export function fetchGcomDashboard(id: string): ThunkResult<void> {
 
 export function importDashboardJson(dashboard: any): ThunkResult<void> {
   return async (dispatch) => {
+    await dispatch(processElements(dashboard));
+    await dispatch(processJsonDashboard(dashboard));
+    dispatch(processInputs());
+  };
+}
+
+const isDataSourceBeingConsumed = (
+  datasource: string | { uid?: string } | undefined | null,
+  datasourceName: string
+): boolean => {
+  if (!datasource) {
+    return false;
+  }
+
+  return typeof datasource === 'string'
+    ? getDataSourceRefNameFromPlaceholder(datasource) === datasourceName
+    : getDataSourceRefNameFromPlaceholder(datasource?.uid) === datasourceName;
+};
+
+function processDashboard(dashboardJson: DashboardJson, state: ImportDashboardState): DashboardJson {
+  let filteredUsedInputs = dashboardJson.__inputs;
+  if (!!state.inputs.libraryPanels.length) {
+    filteredUsedInputs = dashboardJson.__inputs?.filter((input: any) => {
+      if (input.type !== InputType.DataSource) {
+        return true;
+      }
+
+      const isDataSourceInputUsedInAnyLibPanel = state.inputs.libraryPanels?.some((libPanel) =>
+        isDataSourceBeingConsumed(libPanel.model.model.datasource, input.name)
+      );
+      const isDataSourceInputUsedInAnyPanel = dashboardJson.panels?.some((el) => {
+        if (!('datasource' in el)) {
+          return false;
+        }
+        return isDataSourceBeingConsumed(el.datasource, input.name);
+      });
+      const isDataSourceInputUsedInAnyTemplating = dashboardJson.templating?.list?.some((temp) =>
+        isDataSourceBeingConsumed(temp.datasource, input.name)
+      );
+      const isDataSourceInputUsedInAnyAnnotation = dashboardJson.annotations?.list?.some((anno) =>
+        isDataSourceBeingConsumed(anno.datasource, input.name)
```

Comment:
I don't like it at all. Any other suggestion for filtering those datasources that are not being used for the library panel? 
Would be really good to add a property in the datasource/element and indicate that is related to the library panel. Then, filter remove it if the library panel already exists. 
Otherwise, we're forced to do this and also check for other elements like panels, annotations and template variables (and I hope not be missing any other)

Reply:
however, if any other element like a panel for example, is using the same datasource, the input will be just one (that's how it's right now and only will affect the panel). If adding a new property to the datasource input is a good option, we will have to duplicate the input in the json when the json is being created and, if we're importing the dashboard and the panel is a new one, show 2 different inputs (one for the panel and the other one for the lib panel)

Reply:
however, if any other element like a panel for example, is using the same datasource, the input will be just one (that's how it's right now and only will affect the panel). If adding a new property to the datasource input is a good option, we will have to duplicate the input in the json when the json is being created and, if we're importing the dashboard and the panel is a new one, show 2 different inputs (one for the panel and the other one for the lib panel)

Reply:
something like this
```
 "__inputs": [
    {
      "name": "DS_GDEV-PROMETHEUS",
      "label": "gdev-prometheus",
      "description": "",
      "type": "datasource",
      "pluginId": "prometheus",
      "pluginName": "Prometheus"
    },
    {
      "name": "DS_GDEV-PROMETHEUS_FOR_LIBRARY_PANEL",
      "label": "gdev-prometheus-for-library-panel",
      "description": "",
      "type": "datasource",
      "pluginId": "prometheus",
      "pluginName": "Prometheus",
      "usage": "library-panel"
    },
    {
      "name": "DS_GDEV-MYSQL",
      "label": "gdev-mysql",
      "description": "",
      "type": "datasource",
      "pluginId": "mysql",
      "pluginName": "MySQL"
    }
  ],
```

Reply:
I think this last solution is better but how much can I change the json structure? I don't know if it can impact in any other place. Moreover, we would need to also change the gcom jsons. How is it possible?

Reply:
I opted for this approach:

```
{
      "name": "DS_GDEV-PROMETHEUS-FOR-LIBRARY-PANEL",
      "label": "gdev-prometheus",
      "description": "",
      "type": "datasource",
      "pluginId": "prometheus",
      "pluginName": "Prometheus",
      "usage": {
        "libraryPanels": [
          {
            "name": "another prom lib panel",
            "uid": "a0379b21-fa20-4313-bf12-d7fd7ceb6f90"
          },
          {
            "name": "Testing lib panel",
            "uid": "c46a6b49-de40-43b3-982c-1b5e1ec084a4"
          }
        ]
      }
    }
```